File: Python3/LeetCode/0023.合并-k-个升序链表.py
#
# @lc app=leetcode.cn id=23 lang=python3
#
# [23] 合并 K 个升序链表
#
# https://leetcode.cn/problems/merge-k-sorted-lists/description/
#
# algorithms
# Hard (63.10%)
# Likes:    3111
# Dislikes: 0
# Total Accepted:    1.1M
# Total Submissions: 1.8M
# Testcase Example:  '[[1,4,5],[1,3,4],[2,6]]'
#
# 给你一个链表数组，每个链表都已经按升序排列。
#
# 请你将所有链表合并到一个升序链表中，返回合并后的链表。
#
#
#
# 示例 1：
#
# 输入：lists = [[1,4,5],[1,3,4],[2,6]]
# 输出：[1,1,2,3,4,4,5,6]
# 解释：链表数组如下：
# [
# ⁠ 1->4->5,
# ⁠ 1->3->4,
# ⁠ 2->6
# ]
# 将它们合并到一个有序链表中得到。
# 1->1->2->3->4->4->5->6
#
#
# 示例 2：
#
# 输入：lists = []
# 输出：[]
#
#
# 示例 3：
#
# 输入：lists = [[]]
# 输出：[]
#
#
#
#
# 提示：
#
#
# k == lists.length
# 0 <= k <= 10^4
# 0 <= lists[i].length <= 500
# -10^4 <= lists[i][j] <= 10^4
# lists[i] 按 升序 排列
# lists[i].length 的总和不超过 10^4
#
#
#


# @lc code=start
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


from typing import List
from typing import Optional


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None
        result = lists[0]
        for i in range(1, len(lists)):
            result = self.mergeList(result, lists[i])
        return result

    def mergeList(self, list1: Optional[ListNode], list2: Optional[ListNode]):
        if not list1:
            return list2
        elif not list2:
            return list1
        dummy = ListNode(0, None)
        cur = dummy
        while list1 and list2:
            if list1.val <= list2.val:
                cur.next = list1
                list1 = list1.next
                cur = cur.next
            else:
                cur.next = list2
                list2 = list2.next
                cur = cur.next
        if list1:
            cur.next = list1
        elif list2:
            cur.next = list2
        return dummy.next


# @lc code=end

File: C++/LeetCode/代码随想录/05_链表/0023.合并K个升序链表.cpp
/*
给你一个链表数组，每个链表都已经按升序排列。
请你将所有链表合并到一个升序链表中，返回合并后的链表。

示例 1：
输入：lists = [[1,4,5],[1,3,4],[2,6]]
输出：[1,1,2,3,4,4,5,6]
解释：链表数组如下：
[
  1->4->5,
  1->3->4,
  2->6
]
将它们合并到一个有序链表中得到。
1->1->2->3->4->4->5->6

示例 2：
输入：lists = []
输出：[]

示例 3：
输入：lists = [[]]
输出：[]
*/
#include "ListNode.h"
class Solution
{
public:
    ListNode *mergeKLists(vector<ListNode *> &lists)
    {
        if (lists.empty())
            return nullptr;
        ListNode *result = lists[0];
        for (int i = 1; i < lists.size(); i += 1)
        {
            result = mergeTwoLists(result, lists[i]);
        }
        return result;
    }
    ListNode *mergeTwoLists(ListNode *list1, ListNode *list2)
    {
        if (list1 == nullptr)
        {
            return list2;
        }
        else if (list2 == nullptr)
        {
            return list1;
        }
        ListNode *dummy = new ListNode(0);
        ListNode *cur = dummy;
        while (list1 && list2)
        {
            if (list1->val <= list2->val)
            {
                cur->next = list1;
                list1 = list1->next;
                cur = cur->next;
            }
            else if (list1->val > list2->val)
            {
                cur->next = list2;
                list2 = list2->next;
                cur = cur->next;
            }
        }
        if (list1)
            cur->next = list1;
        else if (list2)
            cur->next = list2;
        ListNode *result = dummy->next;
        delete dummy; // 删除dummy节点
        return result;
    }
};
int main()
{
    Solution s;
    vector<ListNode *> lists = {createListNode({1, 4, 5}), createListNode({1, 3, 4}), createListNode({2, 6})};
    ListNode *l3 = s.mergeKLists(lists);
    printListNode(l3);
    // 释放合并后的链表内存
    deleteList(l3);

    return 0;
}